content/systems/filecoin_files/piece/_index.md

```diff
@@ -8,20 +8,36 @@ dashboardAudit: 0
 dashboardTests: 0
 ---
 
-# Piece - Part of a file
+# The Filecoin Piece & Data Representation in Filecoin
 ---
 
+The _Filecoin Piece_ is the main _unit of negotiation_ for data that users store on the Filecoin network. The Filecoin Piece is _not a unit of storage_, it is not of a specific size, but is upper-bounded by the size of the _Sector_. A Filecoin Piece can be of any size, but if a Piece is larger than the size of a Sector that the miner supports it has to be split into more Pieces so that each Piece fits into a Sector.
+
 A `Piece` is an object that represents a whole or part of a `File`,
 and is used by `Clients` and `Miners` in `Deals`. `Clients` hire `Miners`
 to store `Pieces`. 
 
-The piece data structure is designed for proving storage of arbitrary
+The Piece data structure is designed for proving storage of arbitrary
 IPLD graphs and client data. This diagram shows the detailed composition
-of a piece and its proving tree, including both full and bandwidth-optimized
-piece data structures.
+of a Piece and its proving tree, including both full and bandwidth-optimized
+Piece data structures.
 
 
 {{< figure src="pieces.png" title="Pieces, Proving Trees, and Piece Data Structures" zoom="true">}}
 
+## Data Representation in the Filecoin Network
+
+It is important to highlight that data submitted to the Filecoin network go through several transformations before they come to the format at which the `StorageProvider` stores it.
+
+1. When a piece of data, or file is submitted to Filecoin (in some raw system format) it is transformed into a _UnixFS DAG style data representation_ (in case it is not in this format already, e.g., from IPFS-based applications). The hash that represents the root of the IPLD DAG of the UnixFS file is the _Payload CID_, which is used in the Retrieval Market. The Payload CID is identical to an IPFS CID.
```

Comment:
I think this might be how Lotus does it when you import something via a CLI, but I don't think it's a strict system requirement that UnixFS get involved with this. For Filecoin Discover we are also importing some arbitrary IPLD blocks for some of the content that won't be interesting when viewed via UnixFS (i.e. the blockchain contents). From a user-perspective I guess mostly they'll see things as UnixFS but the system just needs data that can get bundled up into a CAR file.

Reply:
"identical to an IPFS CID" isn't really correct. It's a CID in the way that IPFS has CIDs but it's very filecoin specific and now useful outside of filecoin. (another thing I outlined in https://github.com/ipfs-shipyard/ipld-explorer/issues/64#issuecomment-663401567 if it helps)

Reply:
> What are the requirements for the data so that it can be bundled in a CAR file? 

(please do not copy this answer into the spec, I am just explaining mechanics)

- On the protocol level car files have no `must` type requirements (for several reasons, a notable one being difficulty of enforcement). Without the presence of some sort of index, a linearized DAG if indistinguishable from random blocks put in a line one after another (the current car spec does not define such an index: you "parse as you go").
In short `.car` files are a mechanism to efficiently transport a collection of blocks.
A **subset** of this capability is transport of well formed/complete DAGs ( multiple roots, all blocks present ).
A **sub-subset** of that is transport of single-root, deterministically processed DAGs ( you start from the root, and walk the dag left-hand-first, depth-first, skipping any already seen blocks ). You will see such car files being referred to as *"deterministic car"*.

A key implementation detail linking above and below bodies of text: on the **sector/proving** level the `.car` file is an opaque blob of data. Nothing is concerned with its internal structure. You can derive `commP` for a stream of random data, in the same manner you can derive an `commP` for the perfect `.car` file. What ties the two worlds is the presence of the `payloadCid` ( the single root within the `.car` file ) in the deal info recorded in the FIlecoin blockchain.

- In the context of filecoin itself things are slightly more complicated: you request a retrieval and start paying for it incrementally as you receive data, root-first. This makes it prudent to only accept a flow you can validate: the first block is **the** payload CID, the second block must be linked-to from the first one, the third from the second, etc. This is where the determinism comes back into focus, and becomes a strict requirement. Still: there are two submodes of operation:
   - In the default `import`/`retrieve` mode the `.car` for the deal is indeed formed by employing UnixFSv1, with [non-default parameters](https://github.com/filecoin-project/lotus/blob/v0.4.2/build/params_shared_vals.go#L16-L19) and is interpreted as such on retrieval.
   - By activating the `--car` option flag on both calls you can, however, operate over deterministically-ordered `.car` files, **containing a single-root DAG of any encoding and content**.

I believe this about sums things up.

Reply:
> have no must type requirements

well .... they _must_ be IPLD blocks, even if they are just `raw` bytes with an associated CID. That's important to communicate but maybe it's covered enough by linking to the CAR spec.

_(Warning: the following is my basic understanding, I'm **far** from expert in this and the world may have changed since I grasped even this tenuous understanding!)_ When making an online deal (as opposed to the less common offline ones we're focused on currently), I think graphsync gets involved in this process. Locally you do this process of building a CAR from your blocks to generate a local CommP for your deal, then you make the deal and use graphsync to transfer the IPLD blocks to the miner who then builds the CAR again from those blocks and gets to the same CommP.

In this flow, you'd get "deterministic" from the way that the CAR is built from the graphsync output just by default. I don't know how common these offline deals will be but we're being pretty liberal with our CAR layouts for it because there's no need for the CAR to be built the same on both sides, we're building them and then shipping them to the miner as they are.

There are implications for retrieval as @ribasushi said, I don't really understand those. I'm also not sure what the focus of this doc should be and whether it should cover both of these cases. I suspect online deals will be the most common. The client still needs to generate CommP from their data. I suppose a client like lotus will handle all of this process for you and it'll end up being responsible for building the graph of your data—and the most common will be file data which lotus will handle the UnixFSification for you.

wrt CAR files being an opaque blob—I don't know if you can avoid them entirely for anything but offline deals, but maybe that's worth noting too?